HPCC-22312 Failing Std.Date unit tests on macOS

dcamper opened the pull request:
This change specifically involves the %Y formatting directive for year output (not parsing).

Complete list of Date functions that were affected:

DateToString
SecondsToString
TimestampToString
ConvertDateFormat
ConvertDateFormatMultiple

Also included with this change is a new __OS__ ECL constant value:  'linux-macos'.

Signed-off-by: Dan S. Camper <dan.camper@lexisnexisrisk.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [x] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

Manual testing plus exercising (new and modified) tests in TestStd.Date.TestFormat.

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

dcamper commented:
@richardkchapman @ghalliday Please review.

HPCCSmoketest commented:
Process of PR-12709, label: hpcc-22312-datetostring-padded-years starts now.
The reason of this test is: Did not build yet.
Commit ID: cca98278d02197e1d94d6e71ec0134b9937b9d18
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: cca98278d02197e1d94d6e71ec0134b9937b9d18
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |113 |113 |0 |0 |0 |35 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |24 sec  (00:00:24) |
|setup (thor) |11 |11 |0 |48 sec  (00:00:48) |
|setup (roxie) |11 |11 |0 |25 sec  (00:00:25) |
|test (hthor) |840 |840 |0 |184 sec  (00:03:04) |
|test (thor) |763 |763 |0 |683 sec  (00:11:23) |
|test (roxie) |916 |916 |0 |249 sec  (00:04:09) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   29 sec (00:00:29) |  212 sec (00:03:32) |    0 sec (00:00:00) |    2 sec (00:00:02) |   19 sec (00:00:19) | 1431 sec (00:23:51) |   18 sec (00:00:18) | 1711 sec (00:28:31) |

ghalliday commented on ecl/hql/hqllex.l:
```diff
@@ -1044,6 +1044,8 @@ __OS__              {
 #ifdef _WIN32
                         returnToken.setExpr(createConstant("windows"));
+#elif defined(__APPLE__)
+                        returnToken.setExpr(createConstant("linux-macos"));
```
@richardkchapman is this a sensible value?

richardkchapman replied:
Not really. MacOS is not linux or linux-based, though it shares some similarity. I don't see how returning this value helps any with back-compatibility - existing code is still going to have to change - so we may as well get it right.

ghalliday commented on ecllibrary/std/Date.ecl:
```diff
@@ -56,6 +56,47 @@ END;
+ * Windows:	0011
+ * Linux:	11
+ *
+ * All support Linux distributions have glibc, which provides support for
```
What happens if the system is compiled using clang?

dcamper replied:
Good question.  I installed Clang 7.0 on my Ubuntu 18.04 system and used (basically) the same code I used on Windows for testing (which is just spitting out the formatted year value from a valid struct tm).  The result was identical to gcc's result:  Using %Y, small years were returned without leading zeroes, and the precision specifier was supported (so %04Y did pad the result).

TL;DR version:  Clang appears to work the same as gcc, which makes sense since it tries to emulate the Gnu toolchain.  Apple may have modified their version of LLVM/Clang so that the leading zeroes are included.

ghalliday reviewed (commented):
@richardkchapman added a couple of questions

richardkchapman commented:
Has this been tried on Windows?

I'm not at all convinced that this is moving in the right direction, nor that it won't cause major compatibility headaches for existing code (99.99% of it running on Linux) that was relying on the previous behaviour. I suspect we should be changing MacOS implementation to return 2 digit dates (or marking as wontfix, if that's too hard)

dcamper commented:
It has been tested on Windows.

Richard, I agree that this feels like a heavy-handed fix.  It does normalize the returned year value, but what changes affects the majority of our actual installations and it is bit hackish on top of that.  It does not feel worth it.

However, rather than just marking the issue as Won't Fix leaves the Std.Date ECL unit tests failing on macOS.

What do you think about changing this PR so that:

1. We add a (more reasonable) __OS__ value for macOS
2. Use __OS__ to make OS-level unit tests for those that will fail due to this year inconsistency, so they will always pass on all supported platforms
3. Perhaps document the inconsistency, if it is deemed important enough

Thoughts?

dcamper committed:
HPCC-22312 Failing Std.Date unit tests on macOS

When converting dates to strings and using the %Y formatting directive to output the year, macOS pads small years with leading zeroes where other supported operating systems do not.  Rather than force a common output, simply account for the difference in the unit tests.

Signed-off-by: Dan S. Camper <dan.camper@lexisnexisrisk.com>

dcamper commented:
Note:  If this PR is accepted then documentation for __OS__ in the language reference manual will need to be updated.

HPCCSmoketest commented:
Process of PR-12709, label: hpcc-22312-datetostring-padded-years starts now.
The reason of this test is: New commit.
Commit ID: e84be38abaf1a506c4aa6b1c11b48d8eb24206db
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: e84be38abaf1a506c4aa6b1c11b48d8eb24206db
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |113 |113 |0 |0 |0 |34 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |3 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |23 sec  (00:00:23) |
|setup (thor) |11 |11 |0 |43 sec  (00:00:43) |
|setup (roxie) |11 |11 |0 |20 sec  (00:00:20) |
|test (hthor) |841 |841 |0 |206 sec  (00:03:26) |
|test (thor) |764 |764 |0 |677 sec  (00:11:17) |
|test (roxie) |917 |917 |0 |242 sec  (00:04:02) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   19 sec (00:00:19) |  215 sec (00:03:35) |    0 sec (00:00:00) |    2 sec (00:00:02) |   18 sec (00:00:18) | 1454 sec (00:24:14) |   19 sec (00:00:19) | 1727 sec (00:28:47) |

richardkchapman commented:
There is also thorlib.os() that should perhaps be changed? Can be a separate Jira though
